crates/turbo-tasks-fs/src/rope.rs

```diff
@@ -0,0 +1,454 @@
+        self.finish();
+        Rope {
+            length: self.length,
+            data: self.committed.into(),
+        }
+    }
+}
+
+impl From<&'static str> for RopeBuilder {
+    default fn from(bytes: &'static str) -> Self {
+        let mut r = RopeBuilder::default();
+        r.push_static_bytes(bytes.as_bytes());
+        r
+    }
+}
+
+impl From<Vec<u8>> for RopeBuilder {
+    fn from(bytes: Vec<u8>) -> Self {
+        RopeBuilder {
+            length: bytes.len(),
+            committed: vec![],
+            writable: bytes,
+        }
+    }
+}
+
+impl Write for RopeBuilder {
+    fn write(&mut self, bytes: &[u8]) -> IoResult<usize> {
+        self.push_bytes(bytes);
+        Ok(bytes.len())
+    }
+
+    fn flush(&mut self) -> IoResult<()> {
+        self.finish();
+        Ok(())
+    }
+}
+
+impl ops::AddAssign<&'static str> for RopeBuilder {
+    fn add_assign(&mut self, rhs: &'static str) {
+        self.push_static_bytes(rhs.as_bytes());
+    }
+}
+
+impl DeterministicHash for Rope {
+    /// Ropes with similar contents hash the same, regardless of their
+    /// structure.
+    fn deterministic_hash<H: DeterministicHasher>(&self, state: &mut H) {
+        state.write_usize(self.len());
+        self.data.deterministic_hash(state);
+    }
+}
+
+impl Serialize for Rope {
+    /// Ropes are always serialized into contiguous strings, because
+    /// deserialization won't deduplicate and share the Arcs (being the only
+    /// possible owner of a individual "shared" data doesn't make sense).
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        use serde::ser::Error;
+        let s = self.to_str().map_err(Error::custom)?;
+        serializer.serialize_str(&s)
+    }
+}
+
+impl<'de> Deserialize<'de> for Rope {
+    /// Deserializes strings into a contiguous, immutable Rope.
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        let bytes = <Vec<u8>>::deserialize(deserializer)?;
+        Ok(Rope::from(bytes))
+    }
+}
+
+impl DeterministicHash for InnerRope {
+    /// Ropes with similar contents hash the same, regardless of their
+    /// structure. Notice the InnerRope does not contain a length (and any
+    /// shared InnerRopes won't either), so the exact structure isn't
+    /// relevant at this point.
+    fn deterministic_hash<H: DeterministicHasher>(&self, state: &mut H) {
+        for v in self.0.iter() {
+            v.deterministic_hash(state);
+        }
+    }
+}
+
+impl From<Bytes> for InnerRope {
+    fn from(bytes: Bytes) -> Self {
+        InnerRope::from(vec![Local(bytes)])
+    }
+}
+
+impl From<Vec<RopeElem>> for InnerRope {
+    fn from(els: Vec<RopeElem>) -> Self {
+        InnerRope(Arc::new(els))
+    }
+}
+
+impl PartialEq for InnerRope {
+    /// Ropes with similar contents are equals, regardless of their structure.
+    fn eq(&self, other: &Self) -> bool {
+        hash_xxh3_hash64(self) == hash_xxh3_hash64(other)
```

Comment:
Can you explain why? Without ropes, concatenating `"a" + "bc"` and `"ab" + "c"` into a string/vec would output equal strings. Matching that seems useful, a rope isn't its structure but its contents. For reference, Google's Cord implements `==` [similarly](https://github.com/abseil/abseil-cpp/blob/e6044634dd7caec2d79a13aecc9e765023768757/absl/strings/cord.h#L1610-L1616).

If the complaint is the hash of the full contents, I can manually implement the traversal algorithm instead. That would allow us to fast exit.

Reply:
> If the complaint is the hash of the full contents, I can manually implement the traversal algorithm instead. That would allow us to fast exit.

My only complaint is about hashing. yes "a" + "bc" == "ab" + "c" should be true. It internal splitting of a rope shouldn't be visible from outside.